### PR TITLE
Improve gptune package 

### PR DESCRIPTION
**Pull request by liuyangzhuan**
1. Address comments by tldahlgren in PR #26936
2. Add a variant +openmpi to allow installation using non openmpi modules

### CODE REVIEW

File: var/spack/repos/builtin/packages/gptune/package.py
Comment (alalazo) [inline var/spack/repos/builtin/packages/gptune/package.py]: Does this package depend explicitly on `openmpi` only?

File: var/spack/repos/builtin/packages/gptune/package.py
Comment (alalazo) [inline var/spack/repos/builtin/packages/gptune/package.py]: Why a `(build, run)` dependency for `openmpi`? Assuming `openmpi` is the only viable choice, I think this can be written in a more robust way as:
```suggestion
    depends_on('openmpi@4:', when='+openmpi')
    depends_on('mpi', when='+openmpi')
```
since the `mpi` dependency will rule out any other `mpi` provider.

**Review by alalazo — CHANGES_REQUESTED** [review]
A comment and a couple of questions.

File: var/spack/repos/builtin/packages/gptune/package.py
Comment (liuyangzhuan) [reply var/spack/repos/builtin/packages/gptune/package.py]: @alalazo mpi is a mandatory requirement of gptune. 

If the variant +openmpi is enabled, gptune needs openmpi, otherwise, gptune can use mpich, spectrum-mpi, cray-mpich, openmpi or other mpi providers. 

So what's the appropriate way of specifying this?

Comment (liuyangzhuan) [reply var/spack/repos/builtin/packages/gptune/package.py]: But the variant should be named "openmpionly" to avoid confusion.

Comment (liuyangzhuan) [reply var/spack/repos/builtin/packages/gptune/package.py]: I've changed the variant name to "mpispawn"

File: var/spack/repos/builtin/packages/gptune/package.py
Comment (liuyangzhuan) [reply var/spack/repos/builtin/packages/gptune/package.py]: I have changed the variant to "mpispawn" to clarify that openmpi is only needed if one would like to use the mpi spawning based feature in gptune.

**Comment by liuyangzhuan** [comment]
Hi, @alalazo I've made a few changes according to your comments:
1. the variant name is changed from "openmpi" to "mpispawn"
2. I modified hypre package to make it work with gptune. 
3. added variants "superlu" and "hypre"
4. simplified a few lines in the gptune package script

Can you take a look at it again?

**Comment by liuyangzhuan** [comment]
Hi, @adamjstewart, I've made some changes according to Issue #28116. Also there were some previous commits in this PR that address @alalazo comments but haven't been processed. Can you take a look altogether?

**Comment by liuyangzhuan** [comment]
Hi, @adamjstewart, this PR has been open for quite a while. Has anyone in your team been able to look at it? Thanks!

**Comment by adamjstewart** [comment]
@liuyangzhuan I don't have a team, I'm just a volunteer contributor. Maybe someone from @spack/merge-to-develop can review?

**Comment by liuyangzhuan** [comment]
@adamjstewart Thanks for the clarification. Is there a way to join the spack contributors so that I can review PRs myself?

**Comment by adamjstewart** [comment]
That's up to @tgamblin. Note that you can't review and merge your own PRs of course.

**Comment by alalazo** [comment]
@spackbot maintainers

File: var/spack/repos/builtin/packages/hypre/package.py
Comment (alalazo) [inline var/spack/repos/builtin/packages/hypre/package.py]: Is this a debug print?

**Review by alalazo — DISMISSED** [review]
I'm waiting for `hypre` maintainers review (apologies, but the bot didn't ping them). Can you point out in a comment over the patch directive:
1. The provenance of the patch
2. The reason why the patch is needed

?

File: var/spack/repos/builtin/packages/hypre/package.py
Comment (liuyangzhuan) [reply var/spack/repos/builtin/packages/hypre/package.py]: That seems to be a debug print. I don't recall if I added it, or someone else. I think that line can be simply removed. 

@balay Could you take a look at the patch files for hypre I submitted in this PR? 

var/spack/repos/builtin/packages/hypre/ij_gptune.patch
var/spack/repos/builtin/packages/hypre/package.py

It's essentially added a few lines in ij.c to print out the runtime into the runlog, and get the input parameters from command line. Thanks!

Comment (liuyangzhuan) [reply var/spack/repos/builtin/packages/hypre/package.py]: @alalazo I have removed this debug printing line.

**Comment by eugeneswalker** [comment]
We would like to include this in E4S 22.2. Hoping someone on Hypre team can greenlight the changes to hypre so we can merge this ASAP!

**Review by osborn9 — APPROVED** [review]
This looks good to me for the changes to hypre.

**Comment by liuyangzhuan** [comment]
@osborn9  Thanks! @alalazo May we get the ball rolling now?

**Review by eugeneswalker — APPROVED** [review]

**Review by wspear — APPROVED** [review]
Looks good to me.

**Pull request merged**
